frontend/ts-src/login.ts

```diff
@@ -0,0 +1,121 @@
+/* Corna login page. */
+
+
+import { getApiUrl, postData } from "./lib/network.js";
+import {
+    clean,
+    isEmail,
+    spaceAtStart,
+    spaceAtEnd
+} from "./lib/utils.js";
+
+interface LoginData {
+    /* Login schema. */
+    email_address: string;
+    password: string;
+}
+
+
+export function onPageLoad(): void {
+    /* add click event listener.
+    *
+    * @returns { void }
+    */
+
+    const loginButton = document.getElementById("Login") as HTMLButtonElement;
+
+    if (!loginButton) { return; }
+
+    loginButton.addEventListener("click", function(e: UIEvent) {
+        e.preventDefault();
+        parseForm();
+    })
+}
+
+
+export function parseForm(): void {
+    /* Grab all inputs and validate.
+    *
+    * @returns { void }
+    */
+
+    const emailInput = document.getElementById("userEmailInput") as HTMLInputElement;
+    const passwordInput = document.getElementById("userPasswordInput") as HTMLInputElement;
+
+    // these functions will call error handling functions
+    if (!validEmail(emailInput) || !validPassword(passwordInput)) { return; }
+
+    const dataToSend: LoginData = {
+        email_address: clean(emailInput.value),
+        password: passwordInput.value,
+    }
+    postData(dataToSend,"v1/auth/login", errorMessage);
+}
+
+
+export function validEmail(email: HTMLInputElement | null): boolean {
+    /* Validate user email.
+    *
+    * @param { HTMLInputElement | null } email: email to validate
+    * @returns { booloan }
+    */
+
+    // very basic check, server side will do most of the checking
+    const hasErrd: boolean = (
+        !email
+        || !email.value
+        || !isEmail(clean(email.value))
+    );
+
+    if (hasErrd) { errorMessage("Please enter a valid email address"); }
+
+    return !hasErrd
+}
+
+
+export function validPassword(password: HTMLInputElement | null): boolean {
+    /* Validate user password.
+    *
+    * @param { HTMLInputElement | null } password: password to validate
+    * @returns { booloan }
+    */
+
+    const hasErrd: boolean = (
+        !password
+        || !password.value
+        || spaceAtStart(password.value)
+        || spaceAtEnd(password.value)
+    );
+
+    if (hasErrd) {
```

Comment:
Wouldn't this also show if you just don't provide any password at all?

Reply:
yep (it's part of the error check above), no password is also an invalid password.